lib/blocktree/hashtoruntime.go

```diff
@@ -30,10 +32,76 @@ func (h *hashToRuntime) set(hash Hash, instance runtime.Instance) {
 	h.mutex.Lock()
 	defer h.mutex.Unlock()
 	h.mapping[hash] = instance
+	inMemoryRuntimesGauge.Inc()
 }
 
 func (h *hashToRuntime) delete(hash Hash) {
 	h.mutex.Lock()
 	defer h.mutex.Unlock()
 	delete(h.mapping, hash)
+	inMemoryRuntimesGauge.Dec()
+}
+
+func (h *hashToRuntime) hashes() (hashes []common.Hash) {
+	h.mutex.RLock()
+	defer h.mutex.RUnlock()
+
+	return maps.Keys(h.mapping)
+}
+
+// onFinalisation handles pruning and recording on block finalisation.
+// newCanonicalBlockHashes is the block hashes of the blocks newly finalised.
+// The last element is the finalised block hash.
+func (h *hashToRuntime) onFinalisation(newCanonicalBlockHashes []common.Hash) {
+	h.mutex.Lock()
+	defer func() {
+		totalInMemoryRuntimes := len(h.mapping)
+		inMemoryRuntimesGauge.Set(float64(totalInMemoryRuntimes))
+		h.mutex.Unlock()
```

Comment:
Can we update the gauge in an event driven way? We can update the gauge whenever we add/remove from the `h.mapping`.

Reply:
@timwu20 it happens in the methods `(h *hashToRuntime) set` and `(h *hashToRuntime) delete`, we execute this update here because we modify the mapping (remove forks)

Reply:
Maybe move this defer after line 67 then? There are no changes to the gauge if it returns early.